Sources/MessagingPush/PushClickHandler.swift

```diff
@@ -0,0 +1,213 @@
+            targetSelector: #selector(UNUserNotificationCenterDelegate.userNotificationCenter(_:didReceive:withCompletionHandler:)),
+            myClass: PushClickHandlerImpl.self,
+            mySelector: #selector(PushClickHandlerImpl.cio_swizzle_didReceive(_:didReceive:withCompletionHandler:))
+        )
+    }
+
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse) -> CustomerIOParsedPushPayload? {
+        guard let parsedPush = CustomerIOParsedPushPayload.parse(response: response, jsonAdapter: jsonAdapter) else {
+            // push not sent from CIO. exit early
+            return nil
+        }
+
+        if response.didClickOnPush {
+            pushClicked(response, parsedPush: parsedPush)
+        }
+
+        return parsedPush
+    }
+
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) -> Bool {
+        guard let parsedPush = CustomerIOParsedPushPayload.parse(response: response, jsonAdapter: jsonAdapter) else {
+            // push did not come from CIO
+            // Do not call completionHandler() because push did not come from CIO. Another service might have sent it so
+            // allow another SDK
+            // to call the completionHandler()
+            return false
+        }
+
+        if response.didClickOnPush {
+            pushClicked(response, parsedPush: parsedPush)
+        }
+
+        completionHandler()
+        return true
+    }
+
+    private func pushClicked(_ response: UNNotificationResponse, parsedPush: CustomerIOParsedPushPayload) {
+        // TODO: prevent duplicate push metrics and deep link handling.
+
+        // Now we are ready to handle the push click.
+        // Track metrics
+        if sdkConfig.autoTrackPushEvents {
+            customerIO?.trackMetric(deliveryID: parsedPush.deliveryId, event: .opened, deviceToken: parsedPush.deviceToken)
+        }
+
+        // Cleanup files on device that were used when the push was displayed. Files are no longer
+        // needed now that the push is no longer shown.
+        cleanupAfterPushInteractedWith(pushContent: parsedPush)
+
+        // Handle deep link, if there is one attached to push.
+        if let deepLinkUrl = parsedPush.deepLink {
+            deepLinkUtil.handleDeepLink(deepLinkUrl)
+        }
+    }
+
+    // There are files that are created just for displaying a rich push. After a push is interacted with, those files
+    // are no longer needed.
+    // This function's job is to cleanup after a push is no longer being displayed.
+    func cleanupAfterPushInteractedWith(pushContent: CustomerIOParsedPushPayload) {
+        pushContent.cioAttachments.forEach { attachment in
+            let localFilePath = attachment.url
+
+            try? FileManager.default.removeItem(at: localFilePath)
+        }
+    }
```

Comment:
This logic was moved here from `Sources/MessagingPush/MessagingPush+AppDelegate.swift` file.